### PR TITLE
Added support for Agent config, Agents and Agent Groups filtering, sorting and pagination, and bulk operations.

### PR DESCRIPTION
**Pull request by pierresc**
* Implemented Agent Exclusions.
* Added agent config.
* Update Agents resource and add filtering, sorting and pagination support
* Update Agent groups api: add filtering, sorting, and pagination support

### CODE REVIEW

**Comment by tenableAndrew** [comment]
A number (possibly all) new tests here appear to be failing in the Jenkins build. 
```
:test

com.tenable.io.api.BulkAgentApiClientTest > testBulkRemoveAgent FAILED
    java.lang.RuntimeException at BulkAgentApiClientTest.java:89
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at BulkAgentApiClientTest.java:89

com.tenable.io.api.BulkAgentApiClientTest > testBulkAddAgent FAILED
    java.lang.RuntimeException at BulkAgentApiClientTest.java:89
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at BulkAgentApiClientTest.java:89

com.tenable.io.api.BulkAgentApiClientTest > testBulkAgentGroupStatus FAILED
    java.lang.RuntimeException at BulkAgentApiClientTest.java:89
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at BulkAgentApiClientTest.java:89

com.tenable.io.api.BulkAgentApiClientTest > testBulkAgentStatus FAILED
    java.lang.RuntimeException at BulkAgentApiClientTest.java:89
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at BulkAgentApiClientTest.java:89

com.tenable.io.api.BulkAgentApiClientTest > testBulkUnlinkAgent FAILED
    java.lang.RuntimeException at BulkAgentApiClientTest.java:89
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at BulkAgentApiClientTest.java:89

com.tenable.io.api.AgentConfigApiClientTest > testAgentConfigDetails FAILED
    java.lang.AssertionError at AgentConfigApiClientTest.java:51

com.tenable.io.api.AgentConfigApiClientTest > testAgentConfigEdit FAILED
    java.lang.AssertionError at AgentConfigApiClientTest.java:51

com.tenable.io.api.AgentGroupsApiClientTest > testAgentGroups FAILED
    java.lang.RuntimeException at AgentGroupsApiClientTest.java:57
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at AgentGroupsApiClientTest.java:57

com.tenable.io.api.AgentsApiClientTest > testFilters FAILED
    java.lang.RuntimeException at AgentsApiClientTest.java:31
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at AgentsApiClientTest.java:31

com.tenable.io.api.AgentsApiClientTest > testSort FAILED
    java.lang.RuntimeException at AgentsApiClientTest.java:101
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at AgentsApiClientTest.java:101

com.tenable.io.api.AgentsApiClientTest > getAgents FAILED
    java.lang.RuntimeException at AgentsApiClientTest.java:31
        Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException at AgentsApiClientTest.java:31

80 tests completed, 11 failed
:test FAILED

```

I haven't had a check to look into the root cause of the errors. The same number seems occurs whether or not agents exist on the test account.

File: src/test/java/com/tenable/io/api/ScannersApiClientTest.java [hunk 1]
```diff
@@ -67,20 +69,31 @@ public void testGetScans() throws Exception {
...
+    @Ignore("CI-16726")
     @Test
     public void testControlScans() throws Exception {
         TenableIoClient apiClient = new TenableIoClient();
         List<Scanner> scanners = apiClient.getScannersApi().list();
         assertNotNull( scanners );
 
-        ScanResult newScan = createScan( apiClient, scanners.get( 0 ).getId() );
+        Scanner scanner = scanners.get( 0 );
+        for (Scanner scannerItem : scanners) {
```
Comment (Ericky14) [inline src/test/java/com/tenable/io/api/ScannersApiClientTest.java]: Spaces inside parenthesis

File: src/test/java/com/tenable/io/api/ScansApiClientTest.java [hunk 2]
```diff
@@ -338,11 +332,18 @@ private ScanResult createScan( TenableIoClient apiClient, int folderId ) throws
...
+    private int getTestScannerId( TenableIoClient apiClient ) throws TenableIoException {
         List<Scanner> scanners = apiClient.getScannersApi().list();
-        Random rnd = new Random();
-        int index = rnd.nextInt( scanners.size() );
-        return scanners.get( index ).getId();
+
+        Scanner scanner = scanners.get( 0 );
+        for (Scanner scannerItem : scanners) {
+            scanner = scannerItem;
+			if (  scannerItem.getName().equals("US Cloud Scanner")  ) {
```
Comment (Ericky14) [inline src/test/java/com/tenable/io/api/ScansApiClientTest.java]: Indentation weird here.

File: src/test/java/com/tenable/io/api/ScansApiClientTest.java [hunk 3]
```diff
@@ -338,11 +332,18 @@ private ScanResult createScan( TenableIoClient apiClient, int folderId ) throws
...
 
-    private int getRandomScannerId( TenableIoClient apiClient ) throws TenableIoException {
+    private int getTestScannerId( TenableIoClient apiClient ) throws TenableIoException {
         List<Scanner> scanners = apiClient.getScannersApi().list();
-        Random rnd = new Random();
-        int index = rnd.nextInt( scanners.size() );
-        return scanners.get( index ).getId();
+
+        Scanner scanner = scanners.get( 0 );
+        for (Scanner scannerItem : scanners) {
```
Comment (Ericky14) [inline src/test/java/com/tenable/io/api/ScansApiClientTest.java]: Space inside parenthesis

File: src/test/java/com/tenable/io/api/TestBase.java [hunk 4]
```diff
@@ -280,5 +285,21 @@ protected String getNewTestScannerGroupName() {
     private String getNewTestName( String prefix ) {
         return prefix + java.util.UUID.randomUUID().toString().substring( 0, 6 );
     }
+
+
+    protected ScanStatus waitForStatus( TenableIoClient apiClient, int scanId, ScanStatus status ) throws TenableIoException, InterruptedException {
+        ScanDetails details = apiClient.getScansApi().details( scanId );
+        ScanStatus curStatus = details.getInfo().getStatus();
+        while( curStatus != status ) {
```
Comment (Ericky14) [inline src/test/java/com/tenable/io/api/TestBase.java]: Space after `while` and `if` keywords

File: src/test/java/com/tenable/io/api/BulkAgentApiClientTest.java [hunk 5]
```diff
@@ -91,7 +91,7 @@ private int getAgentGroupId() throws Exception {
 
         List<Integer> agentIds = new ArrayList<>();
 
-        for ( int i = 0; i < agents.size(); i++ ) {
+        for ( int i = 2; i < agents.size(); i++ ) {
```
Comment (Ericky14) [inline src/test/java/com/tenable/io/api/BulkAgentApiClientTest.java]: Maybe we could add a case where agentIds returns array less than 2, just to be sure the test won't just crash here. Or assert that the length is greater than 2.

**Pull request merged**